tests/lib/rule-tester/rule-tester.js

```diff
@@ -2295,6 +2296,123 @@ describe("RuleTester", () => {
+            const ruleWithNoMeta = {
+                create(context) {
+                    return {
+                        Program(node) {
+                            context.report({ node, message: "bad" });
+                        }
+                    };
+                }
+            };
+
+            ruleTester.run("rule-with-no-meta", ruleWithNoMeta, {
+                valid: [],
+                invalid: [
+                    { code: "var foo = bar;", options: [{ foo: true }], errors: 1 }
+                ]
+            });
+
+            assert.strictEqual(spy.callCount, 1, "calls `util.deprecate` once");
+            assert.strictEqual(
+                spy.getCall(0).args[1],
+                "\"rule-with-no-meta\" rule has options but is missing the \"meta.schema\" property and will stop working in ESLint v9. Please add a schema: https://eslint.org/docs/developer-guide/working-with-rules#options-schemas"
+            );
+        });
+
+        it("should log a deprecation warning when schema is not defined for the rule", () => {
+            const ruleWithNoSchema = {
+                meta: {
+                    type: "suggestion"
+                },
+                create(context) {
+                    return {
+                        Program(node) {
+                            context.report({ node, message: "bad" });
+                        }
+                    };
+                }
+            };
+
+            ruleTester.run("rule-with-no-schema", ruleWithNoSchema, {
+                valid: [],
+                invalid: [
+                    { code: "var foo = bar;", options: [{ foo: true }], errors: 1 }
+                ]
+            });
+
+            assert.strictEqual(spy.callCount, 1, "calls `util.deprecate` once");
+            assert.strictEqual(
+                spy.getCall(0).args[1],
+                "\"rule-with-no-schema\" rule has options but is missing the \"meta.schema\" property and will stop working in ESLint v9. Please add a schema: https://eslint.org/docs/developer-guide/working-with-rules#options-schemas"
+            );
+        });
+
+        it("should not log a deprecation warning when schema is an empty array", () => {
+            const ruleWithEmptySchema = {
+                meta: {
+                    type: "suggestion",
+                    schema: []
+                },
+                create(context) {
+                    return {
+                        Program(node) {
+                            context.report({ node, message: "bad" });
+                        }
+                    };
+                }
+            };
+
+            ruleTester.run("rule-with-no-options", ruleWithEmptySchema, {
+                valid: [],
+                invalid: [{ code: "var foo = bar;", errors: 1 }]
+            });
+
+            assert.strictEqual(spy.callCount, 0, "never calls `util.deprecate`");
+        });
+    });
```

Comment:
It would be nice to add tests for more cases where the warnings should not be emitted:

* When the rule is an object-style rule, the legacy rule API warning is not emitted.
* When the rule has `meta.schema` and there are test cases with `options`, the missing schema warning is not emitted.
* When the rule does not have `meta`, but there are no test cases with `options`, the missing schema warning is not emitted.
* When the rule has `meta` without `meta.schema`, but there are no test cases with `options`, the missing schema warning is not emitted.
* When the rule has `meta` without `meta.schema`, and some test cases have `options` property but it's an empty array, the missing schema warning is not emitted.

Reply:
I will add these cases 👍🏻